autosign.bash

```diff
@@ -134,30 +210,51 @@ sign_key() {
 	local sign_uid=${2}
 	local ret=1
 
+	local f=sign-key-${key}.txt
+
+	"${gpgcmd[@]}" \
+		"${trust_args[@]}" \
+		--with-colons \
+		--list-keys "${key}" \
+		2>/dev/null >"${f}"
+
 	# verify whether the key is suitable for signing
 	local l trust uid email uids=() need_full=0
 	while read l; do
-		case ${l} in
-			pub:[er]:*)
+		# split the gpg line correctly into fields
+		# Per GPG DETAILS, there will never be a raw : inside a field, it will instead be encoded as \x3a
+		# TODO: as a seperate pass, apply a C decode to convert the above back.
+		IFS=':' read -ra GPGREC <<< "$l"
+
+		# Be a terrible programmer: the docs have fields numbered from 1
+		# So shift up the entire array by one position
+		i=${#GPGREC[@]}
+		while [[ $i -ge 0 ]]; do
+			GPGREC[i+1]="${GPGREC[i]}"
+			let i=i-1
+		done
+		GPGREC[0]='INVALID'
+
+		case ${GPGREC[1]} in
+			pub)
 				# skip expired key
-				return 1
+				# skip revoked key
+				[[ ${GPGREC[2]} =~ [er] ]] && return 1
 				;;
-			uid:*)
-				trust=${l#uid:}
-				uid=${trust#?::::*::*::}
-				trust=${trust%%:*}
-				uid=${uid%%:*}
+			uid)
+				trust=${GPGREC[2]}
+				uid=${GPGREC[10]}
 				email=${uid#*<}
 				email=${email%%>*}
 				[[ -n ${trust} && -n ${email} ]] ||
 					die "Unable to parse uid: ${l}"
-				[[ ${email,,} == ${sign_uid,,} && ${trust} != [er] ]] &&
+				[[ "${email,,}" == "${sign_uid,,}" && ${trust} != [er] ]] &&
 					uids+=( "=${uid}" )
 				# if there are revoked UIDs, they may collide
-				[[ ${trust} == [er] ]] && need_full=1
+				[[ ${trust} =~ [er] ]] && need_full=1
```

Comment:
Why are you using regexps when pattern matching can do the same more readably and faster?